base/parse.jl

```diff
@@ -217,6 +217,69 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
 end
 
 ## string to float functions ##
+function parsefloat_preamble(s::AbstractString, base::Int, i::Int=1)
+    isempty(s) && throw(ArgumentError("Empty string!"))
+    c = ' '
+    sign = 1
+    while isspace(c)
+        c, i = next(s,i)
+    end
+    if c in ('-', '+')
+        sign = (c == '-') ? -1 : 1
+        c, i = next(s, i)
+    end
+    while isspace(c)
+        c, i = next(s,i)
+    end
+    return sign, c, i
+end
+
+function parse{T<:AbstractFloat}(::Type{T}, s::AbstractString, base::Integer)
+    sign, c, i = parsefloat_preamble(s, base)
+
+    lowercase(c) == 'n' && return T(NaN)
+    lowercase(c) == 'i' && return sign*T(Inf)
+
+    res = zero(T)
+    b = T(base)
+    exponent = 0
+    tmpexp = 0
+    while true
+        isspace(c) && break
+        c == '.' && break
+        d = parse(Int, c, base)
+        res = res*b + d
+        done(s, i) && break
+        c, i = next(s, i)
+    end
+
+    n = -1
+    while !done(s, i)
+        c, i = next(s, i)
+        isspace(c) && break
+        c == 'e' && break
```

Comment:
`e` may be a valid digit for e.g. a hex number. Maybe we should allow scientific notation only for bases smaller then `15`, where only digits `0,1,...a,b,c,d` can occur.